### PR TITLE
Compact scanning results and add post-hook to deliver them

### PR DESCRIPTION
**Pull request by mfojtik**
This is a prototype for the compact scanning results discussed here: https://github.com/openshift/image-inspector/issues/49

Invoke image-inspector with `-post-results-url`
```
# image-inspector -image registry.access.redhat.com/rhel7/rsyslog:latest -scan-type openscap -openscap-html-report -post-results-url http://localhost:9090/
2017/06/01 11:40:37 Pulling image registry.access.redhat.com/rhel7/rsyslog:latest
2017/06/01 11:40:41 Extracting image registry.access.redhat.com/rhel7/rsyslog:latest to /var/tmp/image-inspector-125452698
2017/06/01 11:40:43 OpenSCAP scanning /var/tmp/image-inspector-125452698. Placing results in /var/tmp/image-inspector-scan-results-690611761
2017/06/01 11:40:50 Posting results to "http://localhost:9090/" ...
2017/06/01 11:40:50 DEBUG: Success: &{200 OK 200 HTTP/1.1 1 1 map[Date:[Thu, 01 Jun 2017 09:40:50 GMT] Content-Length:[0] Content-Type:[text/plain; charset=utf-8]] 0x8cc210 0 [] false false map[] 0xc4200d05a0 <nil>}
```

Receive the results on the server side:
```
[@dev] ~ # go run server.go
2017/06/01 11:40:50 body: {"imageName":"registry.access.redhat.com/rhel7/rsyslog:latest","imageID":"sha256:7db3fd3db40f7593fdba321e36aa6e65743ef18b68582221b58af762a03f98f8","results":[{"name":"openscap","scannerVersion":"1.2.3","timestamp":"2017-06-01T11:40:50.345604311+02:00","reference":"https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2017-7502","description":"RHSA-2017:1365: nss security and bug fix update (Important)","summary":[{"Label":"high"}]}]}
```

### CODE REVIEW

**Comment by mfojtik** [comment]
@simon3z @pweil- FYI

File: pkg/inspector/image-inspector.go [hunk 1]
```diff
@@ -133,12 +142,32 @@ func (i *defaultImageInspector) Inspect() error {
 		} else {
 			i.meta.OpenSCAP.Status = iiapi.StatusSuccess
 		}
+		scanResults.Results = append(scanResults.Results, openscap.ParseResults(scanReport)...)
```
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: @pweil- all we need here is `clamav.ParseResults()` that converts whatever clam provides to the result format.

File: pkg/openscap/openscap.go [hunk 2]
```diff
@@ -220,3 +222,26 @@ func (s *defaultOSCAPScanner) ResultsFileName() string {
 func (s *defaultOSCAPScanner) HTMLResultsFileName() string {
 	return path.Join(s.ResultsDir, HTMLResultFile)
 }
+
+func ParseResults(report []byte) []iiapi.Result {
+	ret := []iiapi.Result{}
+	node := xmldom.Must(xmldom.ParseXML(string(report))).Root
```
Comment (mfojtik) [inline pkg/openscap/openscap.go]: @pweil- XML will hunt you! :-))

**Comment by simon3z** [comment]
Nice!

@mfojtik any consideration about security for pushing results? E.g. tokens, etc.
Maybe something like `--post-results-token-file=/var/run/secrets/kubernetes.io/...`

**Comment by mfojtik** [comment]
@simon3z in the micro-services world, I would expect there is a service that receives this POST, translate to it other API call with the auth it requires to succeed (depends on the receiver).

I don't really want to go into a business of implementing different authentication mechanisms... What Github does and the openshift webhooks does is that they set a token that both receiver and sender share (so you can protect from fraudolent POST ;-). wdyt?

**Comment by mfojtik** [comment]
also the token can be part of the POST url (http://localhost:9090/foo/bar/TOKEN) in that case only the receiving service and the sender knows it ;-)

**Comment by simon3z** [comment]
> I don't really want to go into a business of implementing different authentication mechanisms... What Github does and the openshift webhooks does is that they set a token that both receiver and sender share (so you can protect from fraudolent POST ;-). wdyt?

@mfojtik that's exactly what I was suggesting.

> also the token can be part of the POST url (http://localhost:9090/foo/bar/TOKEN) in that case only the receiving service and the sender knows it ;-)

@mfojtik problem with that is it's visible in the pod definition (by anyone who has cluster-reader). Using a secret is not much harder and gives improved security.

I think that by having SSL + Token we should be covered, no need for anything else and it would align with all the other services in Kubernetes/OpenShift.

**Comment by mfojtik** [comment]
@simon3z cluster-reader should be given only to admins, right? but yeah, I can add the token, should be easy.

**Comment by mfojtik** [comment]
@simon3z added `-post-results-token-file` option.

**Comment by mfojtik** [comment]
@simon3z @pweil- anything missing here? @aweiteka are you OK with this compact format? Note that this is supplemental to existing openscap reports, just trying to open floor for additional scanning mechanism and make it easier to integrate.

**Comment by pweil-** [comment]
should be able to get time to look this weekend :smile:

**Comment by aweiteka** [comment]
> are you OK with this compact format?

Yes. Output looks good. Post+token looks good too. +1

**Comment by simon3z** [comment]
@enoodle @ilackarms please review

File: pkg/api/types.go [hunk 3]
```diff
@@ -1,8 +1,9 @@
 package api
 
 import (
-	docker "github.com/fsouza/go-dockerclient"
 	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
```
Comment (enoodle) [inline pkg/api/types.go]: why change here? this might cause merge problems..

File: pkg/cmd/types.go [hunk 4]
```diff
@@ -10,6 +10,9 @@ import (
 	"os"
 )
 
+// TODO: Make this configurable?
```
Comment (enoodle) [inline pkg/cmd/types.go]: this is already configurable through "--docker" argument.

File: pkg/inspector/image-inspector.go [hunk 5]
```diff
@@ -133,15 +142,48 @@ func (i *defaultImageInspector) Inspect() error {
...
```
Comment (enoodle) [inline pkg/inspector/image-inspector.go]: Can we call this only if we are going to post the results?

File: pkg/inspector/image-inspector.go [hunk 6]
```diff
@@ -133,15 +142,48 @@ func (i *defaultImageInspector) Inspect() error {
...
 	}
 
 	if i.imageServer != nil {
 		return i.imageServer.ServeImage(&i.meta,
 			scanReport, htmlScanReport)
 	}
+
+	if len(i.opts.PostResultURL) > 0 {
```
Comment (enoodle) [inline pkg/inspector/image-inspector.go]: Can we move this to a function call just for clarity? This function is already too big

File: pkg/openscap/openscap.go [hunk 7]
```diff
@@ -220,3 +222,26 @@ func (s *defaultOSCAPScanner) ResultsFileName() string {
...
+	for _, c := range node.Query("//rule-result") {
+		if !strings.Contains(c.GetChild("result").Text, "fail") {
+			continue
+		}
+		result := iiapi.Result{
+			Name:           "openscap",
+			ScannerVersion: "1.2.3", // FIXME
```
Comment (enoodle) [inline pkg/openscap/openscap.go]: Do we have scanner versions? should this be the output of `oscap --version` ?

File: pkg/openscap/openscap.go [hunk 8]
```diff
@@ -220,3 +222,26 @@ func (s *defaultOSCAPScanner) ResultsFileName() string {
...
+			Timestamp:      time.Now(),
+			Reference:      fmt.Sprintf("https://cve.mitre.org/cgi-bin/cvename.cgi?name=%s", strings.TrimSpace(c.GetChild("ident").Text)),
```
Comment (enoodle) [inline pkg/openscap/openscap.go]: I think we should have this web address in a const

File: pkg/api/types.go [hunk 9]
```diff
@@ -14,6 +15,49 @@ const (
 	StatusError        OpenSCAPStatus = "Error"
 )
 
+// ScanResult represents the compacted result of all scans performed on the image
+type ScanResult struct {
+	// ImageName is a full pull spec of the input image
+	ImageName string `json:"imageName"`
+	// ImageIUD is a SHA256 identifier of the scanned image
+	ImageID string `json:"imageID"`
```
Comment (pweil-) [inline pkg/api/types.go]: Why is this separate than the `ImageName` if `ImageName` is a qualified pull spec?

File: pkg/api/types.go [hunk 10]
```diff
@@ -14,6 +15,49 @@ const (
...
+	// Results contains compacted results of various scans performed on the image.
+	// Empty results means no problems were found with the given image.
+	Results []Result `json:"results,omitempty"`
```
Comment (pweil-) [inline pkg/api/types.go]: Do we need to start thinking about a `Spec` and `Status` separation here for the future?  Use case: where can report that the image-inspector itself failed in a robust way?  Maybe that's just a pod log for now, but it's something I've been considering.

File: pkg/api/types.go [hunk 11]
```diff
@@ -14,6 +15,49 @@ const (
...
+	// Description describes the result in human readable form
+	Description string `json:"description,omitempty"`
+	// Summary contains a list of severities for the given result
+	Summary []Summary `json:"summary,omitempty"`
+}
+
+type Severity string
+
+var (
+	LowSeverity       Severity = "low"
```
Comment (pweil-) [inline pkg/api/types.go]: switch the name so that type is first (`SeverityLow`) so it shows nicely in code complete :-)

File: pkg/api/types.go [hunk 12]
```diff
@@ -14,6 +15,49 @@ const (
...
+	MediumSeverity    Severity = "medium"
+	ImportantSeverity Severity = "important"
```
Comment (pweil-) [inline pkg/api/types.go]: Let's add some comments here.  What is important vs high/medium/low?

File: pkg/api/types.go [hunk 13]
```diff
@@ -14,6 +15,49 @@ const (
...
+	// ScannerVersion is the scanner version
+	ScannerVersion string `json:"scannerVersion"`
+	// Timestamp is the exact time this scan was performed
+	Timestamp time.Time `json:"timestamp"`
+	// Reference contains URL to more details about the given result
+	Reference string `json:"reference"`
+	// Description describes the result in human readable form
+	Description string `json:"description,omitempty"`
+	// Summary contains a list of severities for the given result
+	Summary []Summary `json:"summary,omitempty"`
```
Comment (pweil-) [inline pkg/api/types.go]: I'm not following why summary is a slice?  Wouldn't we want to relate a severity to a description somehow if you have multiple vulnerabilities?  If we're not doing that why wouldn't this be the highest severity found in the scan?

File: pkg/api/types.go (hunk 12 above)
Comment (aweiteka) [inline pkg/api/types.go]: There's this: https://access.redhat.com/security/updates/classification/

* critical
* important
* moderate
* low

Comment (pweil-) [inline pkg/api/types.go]: +1 let's update the descriptions and maybe provide a link too.  Thanks @aweiteka

File: pkg/api/types.go (hunk 3 above)
Comment (mfojtik) [inline pkg/api/types.go]: @enoodle goimports reordered this imports automatically for me (vim-go). what merge problems you have in mind?

File: pkg/api/types.go (hunk 9 above)
Comment (mfojtik) [inline pkg/api/types.go]: The ImageName might be `docker.io/foo/bar:latest`.... I guess we also want to make sure that the image we scanned was the exact image we use in the pod... Actually that opens a new problem for the scanning...
Let say I have Pod that has image set to `docker.io/foo/bar:latest`. The image inspector will be ran with "let scan `docker.io/foo/bar:latest` but that image might not be the same as in PodSpec (might be more new...).

@simon3z how CF deals with this?

File: pkg/api/types.go (hunk 10 above)
Comment (mfojtik) [inline pkg/api/types.go]: I don't think following kube API patterns makes too much sense here (already have been into this discussion with Clayton)... Also I don't think there is any example/reference in Kube for "reports".

File: pkg/api/types.go (hunk 13 above)
Comment (mfojtik) [inline pkg/api/types.go]: @pweil- that is just copying the image quality format.

File: pkg/api/types.go (hunk 11 above)
Comment (mfojtik) [inline pkg/api/types.go]: +1

File: pkg/api/types.go (hunk 9 above)
Comment (enoodle) [inline pkg/api/types.go]: @mfojtik In ManageIQ we are comparing the Id / repo digest of the image [1]

[1]https://github.com/ManageIQ/manageiq-providers-kubernetes/blob/master/app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb#L112

File: pkg/inspector/image-inspector.go (hunk 5 above)
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: @enoodle we also want to serve this via HTTP, see https://github.com/openshift/image-inspector/pull/55

File: pkg/inspector/image-inspector.go (hunk 6 above)
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: sure

File: pkg/openscap/openscap.go (hunk 7 above)
Comment (mfojtik) [inline pkg/openscap/openscap.go]: it should be output of oscap --version, I will add it

Comment (mfojtik) [inline pkg/openscap/openscap.go]: @enoodle actually the reason I didn't added it was this:

```
[@dev] .../openshift/clamav # oscap -q -V | wc -l
92
```

i really hate to parse the version out. how about setting this to '1.2' and making it a const for now?

Comment (enoodle) [inline pkg/openscap/openscap.go]: Ok :+1:

File: pkg/inspector/image-inspector.go (hunk 6 above)
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: (done)

**Comment by mfojtik** [comment]
code comments addressed

**Comment by mfojtik** [comment]
@enoodle @simon3z @pweil- any other comments? will rebase the clam scan type on top of this.

File: pkg/api/types.go (hunk 3 above)
Comment (enoodle) [inline pkg/api/types.go]: I really don't know how it works. For me it leaves them as they were before or orders them this way if I put a line between.. This creates unnecessary merge problems when two PRs are reordering those (already happened here).

**Pull request merged**